src/main/java/org/broadinstitute/hellbender/tools/dataflow/MarkDuplicatesDataflow.java

```diff
@@ -312,6 +391,205 @@ public void processElement(final ProcessContext context) throws Exception {
+                                } else {
+                                    ++metrics.READ_PAIR_DUPLICATES;
+                                }
+                            }
+                            if (record.hasAttribute(OPTICAL_DUPLICATE_ATTRIBUTE_NAME) &&
+                                record.getAttributeAsInteger(OPTICAL_DUPLICATE_ATTRIBUTE_NAME) == 1) {
+                                ++metrics.READ_PAIR_OPTICAL_DUPLICATES;
+                            }
+                            final KV<String, DuplicationMetrics> kv = KV.of(library, metrics);
+                            context.output(kv);
+                        }
+                    });
+        }
+
+
+        public static class CombineMetricsFn implements SerializableFunction<Iterable<DuplicationMetrics>, DuplicationMetrics> {
+            private static final long serialVersionUID = 1l;
+
+            @Override
+            public DuplicationMetrics apply(Iterable<DuplicationMetrics> input) {
+                DuplicationMetrics metricsSum = new DuplicationMetrics();
+                int num = 0;
+                for (final DuplicationMetrics metrics : input) {
+                    if (metricsSum.LIBRARY == null) {
+                        metricsSum.LIBRARY = metrics.LIBRARY;
+                    }
+                    // This should never happen, as we grouped by key using library as the key.
+                    if (!metricsSum.LIBRARY.equals(metrics.LIBRARY)) {
+                        return null;
+                    }
+                    metricsSum.UNMAPPED_READS += metrics.UNMAPPED_READS;
+                    metricsSum.UNPAIRED_READS_EXAMINED += metrics.UNPAIRED_READS_EXAMINED;
+                    metricsSum.READ_PAIRS_EXAMINED += metrics.READ_PAIRS_EXAMINED;
+                    metricsSum.UNPAIRED_READ_DUPLICATES += metrics.UNPAIRED_READ_DUPLICATES;
+                    metricsSum.READ_PAIR_DUPLICATES += metrics.READ_PAIR_DUPLICATES;
+                    metricsSum.READ_PAIR_OPTICAL_DUPLICATES += metrics.READ_PAIR_OPTICAL_DUPLICATES;
+                }
+                metricsSum.PERCENT_DUPLICATION = 0.0;
+                metricsSum.ESTIMATED_LIBRARY_SIZE = 0L;
+                return metricsSum;
+            }
+        }
+
+        private PTransform<PCollection<? extends KV<String, DuplicationMetrics>>, PCollection<KV<String, DuplicationMetrics>>> finalizeMetrics() {
+            return ParDo
+                .named("finalize metrics")
+                .of(new DoFn<KV<String, DuplicationMetrics>, KV<String, DuplicationMetrics>>() {
+                        private static final long serialVersionUID = 1l;
+
+                        @Override
+                        public void processElement(final ProcessContext context) throws Exception {
+                            DuplicationMetrics metrics = context.element().getValue();
+                            metrics.READ_PAIRS_EXAMINED = metrics.READ_PAIRS_EXAMINED / 2;
```

Comment:
could you add a comment about this / 2?  Is it because it's because it's counted for every read when it should be counted for pairs?